mithril-common/src/cardano_block_scanner/chain_reader_block_streamer.rs

```diff
@@ -0,0 +1,320 @@
+                }
+            }
+            Some(ChainBlockNextAction::RollBackward { rollback_point }) => {
+                debug!(
+                    self.logger,
+                    "ChainReaderBlockStreamer received a RollBackward({rollback_point:?})"
+                );
+                let block_streamer_next_action = if rollback_point == self.from {
+                    BlockStreamerNextAction::SkipToNextAction
+                } else {
+                    BlockStreamerNextAction::ChainBlockNextAction(
+                        ChainBlockNextAction::RollBackward { rollback_point },
+                    )
+                };
+                Ok(Some(block_streamer_next_action))
+            }
+            None => {
+                debug!(self.logger, "ChainReaderBlockStreamer received nothing");
+                Ok(None)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cardano_block_scanner::ScannedBlock;
+    use crate::chain_reader::FakeChainReader;
+    use crate::test_utils::TestLogger;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_parse_expected_nothing_above_block_number_threshold() {
+        let logger = TestLogger::stdout();
+        let chain_reader = Arc::new(Mutex::new(FakeChainReader::new(vec![
+            ChainBlockNextAction::RollForward {
+                parsed_block: ScannedBlock::new("hash-1", 1, 10, 20, Vec::<&str>::new()),
+            },
+        ])));
+        let mut block_streamer =
+            ChainReaderBlockStreamer::try_new(chain_reader, None, 1, logger.clone())
+                .await
+                .unwrap();
+
+        let scanned_blocks = block_streamer.poll_next().await.expect("poll_next failed");
+
+        assert_eq!(None, scanned_blocks,);
+    }
+
+    #[tokio::test]
+    async fn test_parse_expected_multiple_rollforwards_below_block_number_threshold() {
+        let logger = TestLogger::stdout();
+        let chain_reader = Arc::new(Mutex::new(FakeChainReader::new(vec![
+            ChainBlockNextAction::RollForward {
+                parsed_block: ScannedBlock::new("hash-1", 1, 10, 1, Vec::<&str>::new()),
+            },
+            ChainBlockNextAction::RollForward {
+                parsed_block: ScannedBlock::new("hash-2", 2, 20, 1, Vec::<&str>::new()),
+            },
+        ])));
+        let mut block_streamer =
+            ChainReaderBlockStreamer::try_new(chain_reader, None, 100, logger.clone())
+                .await
+                .unwrap();
+
+        let scanned_blocks = block_streamer.poll_next().await.expect("poll_next failed");
+
+        assert_eq!(
+            Some(ChainScannedBlocks::RollForwards(vec![
+                ScannedBlock::new("hash-1", 1, 10, 1, Vec::<&str>::new()),
+                ScannedBlock::new("hash-2", 2, 20, 1, Vec::<&str>::new())
+            ])),
+            scanned_blocks,
+        );
+    }
+
+    #[tokio::test]
+    async fn test_parse_expected_maximum_rollforwards_retrieved_per_poll() {
```

Comment:
Executing another `poll_next()` and verifying we are getting the correct value for `scanned_blocks` here or in another test would be relevant?

Reply:
Done